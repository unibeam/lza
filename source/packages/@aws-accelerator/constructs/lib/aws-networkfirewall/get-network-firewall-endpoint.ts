/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as path from 'path';

interface IGetNetworkFirewallEndpoint extends cdk.IResource {
  /**
   * The ID of the endpoint
   */
  readonly endpointId: string;
}

interface GetNetworkFirewallEndpointProps {
  /**
   * The AZ the endpoint is located in
   */
  readonly endpointAz: string;

  /**
   * The ARN of the associated Network Firewall
   */
  readonly firewallArn: string;

  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;

  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;

  /**
   * The region of the Network Firewall
   */
  readonly region: string;
}

export class GetNetworkFirewallEndpoint extends cdk.Resource implements IGetNetworkFirewallEndpoint {
  public readonly endpointId: string;

  constructor(scope: Construct, id: string, props: GetNetworkFirewallEndpointProps) {
    super(scope, id);

    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, 'Custom::GetNetworkFirewallEndpoint', {
      codeDirectory: path.join(__dirname, 'get-network-firewall-endpoint/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements: [
        {
          Effect: 'Allow',
          Action: ['ec2:DescribeAvailabilityZones', 'network-firewall:DescribeFirewall'],
          Resource: '*',
        },
      ],
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: 'Custom::GetNetworkFirewallEndpoint',
      serviceToken: provider.serviceToken,
      properties: {
        endpointAz: props.endpointAz,
        firewallArn: props.firewallArn,
        region: props.region,
      },
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.kmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.endpointId = resource.ref;
  }
}
