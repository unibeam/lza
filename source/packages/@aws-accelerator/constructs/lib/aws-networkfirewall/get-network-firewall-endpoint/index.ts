/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { DescribeAvailabilityZonesCommand, EC2Client } from '@aws-sdk/client-ec2';
import { DescribeFirewallCommand, NetworkFirewallClient } from '@aws-sdk/client-network-firewall';

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
/**
 * get-network-firewall-endpoint - lambda handler
 *
 * @param event
 * @returns
 */

export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string | undefined;
      Status: string;
    }
  | undefined
> {
  // Set variables passed in event
  const endpointAz: string = event.ResourceProperties['endpointAz'];
  const firewallArn: string = event.ResourceProperties['firewallArn'];
  const region: string = event.ResourceProperties['region'];
  const solutionId = process.env['SOLUTION_ID'];

  const ec2Client = new EC2Client({ customUserAgent: solutionId });
  const nfwClient = new NetworkFirewallClient({ customUserAgent: solutionId });

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      const logicalZoneName = await getAvailabilityZone(ec2Client, endpointAz, region);
      let endpointId: string | undefined = undefined;

      try {
        const response = await throttlingBackOff(() =>
          nfwClient.send(new DescribeFirewallCommand({ FirewallArn: firewallArn })),
        );
        //
        // Check for endpoint in specified AZ
        if (response.FirewallStatus?.SyncStates) {
          endpointId = response.FirewallStatus.SyncStates[logicalZoneName].Attachment?.EndpointId;
        }
        //
        // Validate endpoint ID
        if (!endpointId) {
          throw new Error(`Unable to locate Network Firewall endpoint in AZ ${endpointAz}`);
        }

        return {
          PhysicalResourceId: endpointId,
          Status: 'SUCCESS',
        };
      } catch (e) {
        throw new Error(`Error retrieving Network Firewall endpoint: ${e}`);
      }

    case 'Delete':
      // Do nothing
      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}

/**
 * Returns the logical zone name for the specified AZ
 * @param ec2Client
 * @param endpointAz
 * @param region
 * @returns
 */
async function getAvailabilityZone(ec2Client: EC2Client, endpointAz: string, region: string): Promise<string> {
  if (endpointAz.includes(region)) {
    return endpointAz;
  }

  try {
    const response = await throttlingBackOff(() =>
      ec2Client.send(
        new DescribeAvailabilityZonesCommand({
          ZoneIds: [endpointAz],
        }),
      ),
    );
    //
    // Validate response
    if (!response.AvailabilityZones) {
      throw new Error(`Unable to retrieve details for AZ ${endpointAz}`);
    }
    if (!response.AvailabilityZones[0].ZoneName) {
      throw new Error(`Unable to retrieve logical zone name for AZ ${endpointAz}`);
    }
    return response.AvailabilityZones[0].ZoneName;
  } catch (e) {
    throw new Error(`Error retrieving logical zone name: ${e}`);
  }
}
