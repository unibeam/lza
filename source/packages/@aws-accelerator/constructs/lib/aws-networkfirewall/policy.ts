/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { NfwRuleSourceCustomActionConfig } from '@aws-accelerator/config';
import { transformPolicy } from './utils';

interface INetworkFirewallPolicy extends cdk.IResource {
  /**
   * The Amazon Resource Name (ARN) of the policy.
   */
  readonly policyArn: string;

  /**
   * The ID of the policy.
   */
  readonly policyId: string;

  /**
   * The name of the policy.
   */
  readonly policyName: string;
}

interface StatefulRuleGroupReference {
  resourceArn: string;
  priority?: number;
}

interface StatelessRuleGroupReference {
  priority: number;
  resourceArn: string;
}

export interface FirewallPolicyProperty {
  statelessDefaultActions: string[];
  statelessFragmentDefaultActions: string[];
  statefulDefaultActions?: string[];
  statefulEngineOptions?: string;
  statefulRuleGroupReferences?: StatefulRuleGroupReference[];
  statelessCustomActions?: NfwRuleSourceCustomActionConfig[];
  statelessRuleGroupReferences?: StatelessRuleGroupReference[];
}

interface NetworkFirewallPolicyProps {
  /**
   * The traffic filtering behavior of a firewall policy, defined in a collection of stateless and stateful rule groups and other settings.
   */
  readonly firewallPolicy: FirewallPolicyProperty;

  /**
   * The descriptive name of the firewall policy.
   */
  readonly name: string;

  /**
   * A description of the firewall policy.
   */
  readonly description?: string;

  /**
   * An optional list of CloudFormation tags.
   */
  readonly tags?: cdk.CfnTag[];
}

abstract class NetworkFirewallPolicyBase extends cdk.Resource implements INetworkFirewallPolicy {
  public abstract readonly policyArn: string;
  public abstract readonly policyId: string;
  public abstract readonly policyName: string;
}

export class NetworkFirewallPolicy extends NetworkFirewallPolicyBase {
  public readonly policyArn: string;
  public readonly policyId: string;
  public readonly policyName: string;

  /**
   * Returns CfnFirewallPolicy by applying updates to included resource
   * @param scope Stack in which included FirewallPolicy is created/managed
   * @param id logicalId of FirewallPolicy
   * @param attrs
   */
  static includedCfnResource(
    scope: cdk.cloudformation_include.CfnInclude,
    id: string,
    props: NetworkFirewallPolicyProps,
  ) {
    const resource = scope.getResource(id) as cdk.aws_networkfirewall.CfnFirewallPolicy;
    // Transform properties as necessary
    resource.firewallPolicy = transformPolicy(props.firewallPolicy);
    resource.description = props.description;
    return resource;
  }

  static fromAttributes(
    scope: Construct,
    id: string,
    attrs: { policyArn: string; policyName: string },
  ): INetworkFirewallPolicy {
    class Import extends NetworkFirewallPolicyBase {
      public readonly policyArn = attrs.policyArn;
      public readonly policyName = attrs.policyName;
      // policyId is not used anywhere. Need to store in SSM if needed
      public readonly policyId = '';

      constructor(scope: Construct, id: string) {
        super(scope, id);
      }
    }
    return new Import(scope, id);
  }

  constructor(scope: Construct, id: string, props: NetworkFirewallPolicyProps) {
    super(scope, id);

    // Set initial properties
    this.policyName = props.name;

    // Set firewall policy property
    const firewallPolicy = transformPolicy(props.firewallPolicy);

    // Set name tag
    props.tags?.push({ key: 'Name', value: this.policyName });

    const resource = new cdk.aws_networkfirewall.CfnFirewallPolicy(this, 'Resource', {
      firewallPolicy: firewallPolicy,
      firewallPolicyName: this.policyName,
      description: props.description,
      tags: props.tags,
    });

    this.policyArn = resource.ref;
    this.policyId = resource.attrFirewallPolicyId;
  }
}
