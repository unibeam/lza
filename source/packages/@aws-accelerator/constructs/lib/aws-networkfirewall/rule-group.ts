/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { NfwRuleGroupRuleConfig } from '@aws-accelerator/config';
import { transformRuleGroup } from './utils';

interface INetworkFirewallRuleGroup extends cdk.IResource {
  /**
   * The Amazon Resource Name (ARN) of the rule group.
   */
  readonly groupArn: string;

  /**
   * The ID of the rule group.
   */
  readonly groupId: string;

  /**
   * The name of the rule group.
   */
  readonly groupName: string;
}

interface NetworkFirewallRuleGroupProps {
  /**
   * The maximum operating resources that this rule group can use.
   */
  readonly capacity: number;

  /**
   * The name of the rule group.
   */
  readonly name: string;

  /**
   * Indicates whether the rule group is stateless or stateful.
   */
  readonly type: string;

  /**
   * A description of the rule group.
   */
  readonly description?: string;

  /**
   * An object that defines the rule group rules.
   */
  readonly ruleGroup?: NfwRuleGroupRuleConfig;

  /**
   * An optional list of CloudFormation tags.
   */
  readonly tags?: cdk.CfnTag[];
}

export class NetworkFirewallRuleGroup extends cdk.Resource implements INetworkFirewallRuleGroup {
  public readonly groupArn: string;
  public readonly groupId: string;
  public readonly groupName: string;

  /**
   * Returns CfnRuleGroup by applying updates to included resource
   * @param scope Stack in which included RuleGroup is created/managed
   * @param id logicalId of RuleGroup
   * @param attrs
   */
  static includedCfnResource(
    scope: cdk.cloudformation_include.CfnInclude,
    id: string,
    props: NetworkFirewallRuleGroupProps,
  ) {
    const resource = scope.getResource(id) as cdk.aws_networkfirewall.CfnRuleGroup;
    // Transform properties as necessary
    if (props.ruleGroup) {
      // Set rule group property
      resource.ruleGroup = transformRuleGroup(props.ruleGroup);
    } else {
      // Remove existing rule group property
      resource.ruleGroup = undefined;
    }
    // Updating capacity requires replacement
    resource.capacity = props.capacity;
    // Updating type requires replacement
    resource.type = props.type;
    resource.description = props.description;
    return resource;
  }

  static fromAttributes(
    scope: Construct,
    id: string,
    attrs: { groupArn: string; groupName: string },
  ): INetworkFirewallRuleGroup {
    class Import extends cdk.Resource implements INetworkFirewallRuleGroup {
      public readonly groupArn = attrs.groupArn;
      public readonly groupId = attrs.groupName;
      // groupId is not used anywhere. Need to store in SSM if needed
      public readonly groupName = '';

      constructor(scope: Construct, id: string) {
        super(scope, id);
      }
    }
    return new Import(scope, id);
  }

  constructor(scope: Construct, id: string, props: NetworkFirewallRuleGroupProps) {
    super(scope, id);

    // Set initial properties
    this.groupName = props.name;

    let ruleGroup: cdk.aws_networkfirewall.CfnRuleGroup.RuleGroupProperty | undefined;
    // Transform properties as necessary
    if (props.ruleGroup) {
      // Set rule group property
      ruleGroup = transformRuleGroup(props.ruleGroup);
    }

    // Set name tag
    props.tags?.push({ key: 'Name', value: this.groupName });

    const resource = new cdk.aws_networkfirewall.CfnRuleGroup(this, 'Resource', {
      capacity: props.capacity,
      ruleGroupName: this.groupName,
      type: props.type,
      description: props.description,
      ruleGroup,
      tags: props.tags,
    });

    this.groupArn = resource.ref;
    this.groupId = resource.attrRuleGroupId;
  }
}
