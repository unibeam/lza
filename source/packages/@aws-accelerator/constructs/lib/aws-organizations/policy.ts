/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import * as assets from 'aws-cdk-lib/aws-s3-assets';
import { Construct } from 'constructs';
import * as path from 'path';
import { createHash } from 'crypto';
import { readFileSync } from 'fs';
import { v4 as uuidv4 } from 'uuid';
import { createLogger } from '@aws-accelerator/utils/lib/logger';

const logger = createLogger(['constructs-organization-policy']);

export enum PolicyType {
  AISERVICES_OPT_OUT_POLICY = 'AISERVICES_OPT_OUT_POLICY',
  BACKUP_POLICY = 'BACKUP_POLICY',
  SERVICE_CONTROL_POLICY = 'SERVICE_CONTROL_POLICY',
  TAG_POLICY = 'TAG_POLICY',
  CHATBOT_POLICY = 'CHATBOT_POLICY',
}

export interface Tag {
  /**
   * The key identifier, or name, of the tag.
   */
  Key: string | undefined;

  /**
   * The string value that's associated with the key of the tag. You can set the value of a
   * tag to an empty string, but you can't set the value of a tag to null.
   */
  Value: string | undefined;
}

/**
 * Initialized Policy properties
 */
export interface PolicyProps {
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
  /**
   * The friendly name of the policy
   */
  readonly name: string;
  /**
   * The AWS partition the policy will be created in
   */
  readonly partition: string;
  /**
   * The path of the file for the policy
   */
  readonly path: string;
  /**
   * The type of policy to create
   */
  readonly type: PolicyType;
  /**
   * The SCP strategy - "allow-list" or "deny-list"The type of policy to create
   */
  readonly strategy?: string;
  /**
   * Accelerator prefix
   */
  readonly acceleratorPrefix: string;
  /**
   * An optional description of the policy
   */
  readonly description?: string;
  /**
   * An optional list of tags for the policy
   */
  readonly tags?: Tag[];
}

/**
 * Class to initialize Policy
 */
export class Policy extends Construct {
  public readonly id: string;
  public readonly path: string;
  public readonly name: string;
  public readonly description?: string;
  public readonly type: PolicyType;
  public readonly strategy?: string;
  public readonly tags?: Tag[];

  constructor(scope: Construct, id: string, props: PolicyProps) {
    super(scope, id);

    this.path = props.path;
    this.name = props.name;
    this.description = props.description || '';
    this.type = props.type;
    this.strategy = props.strategy;
    this.tags = props.tags || [];

    //
    // Bundle the policy file. This will be available as an asset in S3
    //
    const asset = new assets.Asset(this, 'Policy', {
      path: props.path,
    });

    let uuid: string;
    // generating md5 hash to allow for file changes and only trigger change when file is updated
    const fileHash = createHash('md5').update(readFileSync(props.path)).digest('hex');

    // Boolean to force update
    const forceUpdate = process.env['ACCELERATOR_FORCED_UPDATE']
      ? process.env['ACCELERATOR_FORCED_UPDATE'] === 'true'
      : false;

    if (forceUpdate) {
      logger.warn(`ACCELERATOR_FORCED_UPDATE env variable is set. Forcing an update.`);
      uuid = uuidv4();
    } else {
      uuid = fileHash;
    }

    //
    // Function definition for the custom resource
    //
    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, 'Custom::OrganizationsCreatePolicy', {
      codeDirectory: path.join(__dirname, 'create-policy/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      description: 'Organizations create policy',
      policyStatements: [
        {
          Effect: 'Allow',
          Action: [
            'organizations:CreatePolicy',
            'organizations:DeletePolicy',
            'organizations:DetachPolicy',
            'organizations:ListPolicies',
            'organizations:ListTargetsForPolicy',
            'organizations:UpdatePolicy',
            'organizations:TagResource',
          ],
          Resource: '*',
        },
        {
          Effect: 'Allow',
          Action: ['s3:GetObject'],
          Resource: cdk.Stack.of(this).formatArn({
            service: 's3',
            region: '',
            account: '',
            resource: asset.s3BucketName,
            arnFormat: cdk.ArnFormat.SLASH_RESOURCE_NAME,
            resourceName: '*',
          }),
        },
      ],
    });

    //
    // Custom Resource definition. We want this resource to be evaluated on
    // every CloudFormation update, so we generate a new uuid to force
    // re-evaluation.
    //
    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: 'Custom::CreatePolicy',
      serviceToken: provider.serviceToken,
      properties: {
        bucket: asset.s3BucketName,
        key: asset.s3ObjectKey,
        partition: props.partition,
        policyTagKey: `${props.acceleratorPrefix}Managed`,
        uuid,
        name: props.name,
        description: props.description,
        type: props.type,
        strategy: props.strategy,
        tags: props.tags,
      },
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.kmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.id = resource.ref;
  }
}
