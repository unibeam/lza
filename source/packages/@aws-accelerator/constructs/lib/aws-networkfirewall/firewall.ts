/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { GetNetworkFirewallEndpoint } from './get-network-firewall-endpoint';

export interface INetworkFirewall extends cdk.IResource {
  /**
   * The unique IDs of the firewall endpoints for all of the subnets that you attached to the firewall.
   */
  readonly endpointIds?: string[];
  /**
   * The Amazon Resource Name (ARN) of the firewall.
   */
  readonly firewallArn: string;

  /**
   * The ID of the firewall.
   */
  readonly firewallId?: string;

  /**
   * The name of the policy.
   */
  readonly firewallName: string;

  addLogging: (config: cdk.aws_networkfirewall.CfnLoggingConfiguration.LoggingConfigurationProperty) => void;

  addNetworkFirewallRoute: (
    id: string,
    endpointAz: string,
    logRetentionInDays: number,
    routeTableId: string,
    destination?: string,
    ipv6Destination?: string,
    logGroupKmsKey?: cdk.aws_kms.IKey,
  ) => void;
}

interface NetworkFirewallProps {
  /**
   * The Amazon Resource Name (ARN) of the firewall policy.
   */
  readonly firewallPolicyArn: string;

  /**
   * The descriptive name of the firewall.
   */
  readonly name: string;

  /**
   * The subnets that Network Firewall is using for the firewall.
   */
  readonly subnets: string[];

  /**
   * The unique identifier of the VPC where the firewall is in use.
   */
  readonly vpcId: string;

  /**
   * A flag indicating whether it is possible to delete the firewall.
   */
  readonly deleteProtection?: boolean;

  /**
   * A description of the firewall.
   */
  readonly description?: string;

  /**
   * A setting indicating whether the firewall is protected against a change to the firewall policy association.
   */
  readonly firewallPolicyChangeProtection?: boolean;

  /**
   * A setting indicating whether the firewall is protected against changes to the subnet associations.
   */
  readonly subnetChangeProtection?: boolean;

  /**
   * An optional list of CloudFormation tags.
   */
  readonly tags?: cdk.CfnTag[];
}

abstract class NetworkFirewallBase extends cdk.Resource implements INetworkFirewall {
  public abstract readonly firewallArn: string;
  public abstract readonly firewallName: string;
  public abstract readonly firewallId?: string;
  public abstract readonly endpointIds?: string[];
  // protected abstract subnetMapping?: cdk.aws_networkfirewall.CfnFirewall.SubnetMappingProperty[];

  public addLogging(config: cdk.aws_networkfirewall.CfnLoggingConfiguration.LoggingConfigurationProperty) {
    new cdk.aws_networkfirewall.CfnLoggingConfiguration(this, 'LoggingConfig', {
      firewallArn: this.firewallArn,
      loggingConfiguration: config,
    });
  }

  public addNetworkFirewallRoute(
    id: string,
    endpointAz: string,
    logRetentionInDays: number,
    routeTableId: string,
    destination?: string,
    ipv6Destination?: string,
    logGroupKmsKey?: cdk.aws_kms.IKey,
  ): void {
    // Get endpoint ID from custom resource
    const vpcEndpointId = new GetNetworkFirewallEndpoint(this, `${id}Endpoint`, {
      endpointAz: endpointAz,
      firewallArn: this.firewallArn,
      kmsKey: logGroupKmsKey,
      logRetentionInDays: logRetentionInDays,
      region: cdk.Stack.of(this).region,
    }).endpointId;

    new cdk.aws_ec2.CfnRoute(this, id, {
      routeTableId,
      destinationCidrBlock: destination,
      destinationIpv6CidrBlock: ipv6Destination,
      vpcEndpointId,
    });
  }
}
export class NetworkFirewall extends NetworkFirewallBase {
  public readonly firewallArn: string;
  public readonly firewallName: string;
  public readonly firewallId?: string;
  public readonly endpointIds?: string[];
  protected subnetMapping: cdk.aws_networkfirewall.CfnFirewall.SubnetMappingProperty[];

  static fromAttributes(
    scope: Construct,
    id: string,
    attrs: {
      firewallArn: string;
      firewallName: string;
    },
  ): INetworkFirewall {
    class Import extends NetworkFirewallBase {
      public readonly firewallArn: string;
      public readonly firewallName: string;
      public readonly firewallId?: string;
      public readonly endpointIds?: string[];
      constructor(scope: Construct, id: string) {
        super(scope, id);
        this.firewallArn = attrs.firewallArn;
        this.firewallName = attrs.firewallName;
      }
    }
    return new Import(scope, id);
  }

  /**
   * Returns CfnFirewall by applying updates to included resource
   * @param scope Stack in which included Firewall is created/managed
   * @param id logicalId of Firewall
   * @param attrs
   */
  static includedCfnResource(scope: cdk.cloudformation_include.CfnInclude, id: string, props: NetworkFirewallProps) {
    // Set initial properties
    const subnetMapping = props.subnets.map(item => {
      return { subnetId: item };
    });
    const resource = scope.getResource(id) as cdk.aws_networkfirewall.CfnFirewall;
    resource.deleteProtection = props.deleteProtection;
    resource.subnetChangeProtection = props.subnetChangeProtection;
    resource.subnetMappings = subnetMapping;
    resource.firewallPolicyArn = props.firewallPolicyArn;
    resource.firewallPolicyChangeProtection = props.firewallPolicyChangeProtection;
    resource.description = props.description;
    return resource;
  }

  constructor(scope: Construct, id: string, props: NetworkFirewallProps) {
    super(scope, id);

    // Set initial properties
    this.firewallName = props.name;
    this.subnetMapping = props.subnets.map(item => {
      return { subnetId: item };
    });

    // Set name tag
    props.tags?.push({ key: 'Name', value: this.firewallName });

    const resource = new cdk.aws_networkfirewall.CfnFirewall(this, 'Resource', {
      firewallName: this.firewallName,
      firewallPolicyArn: props.firewallPolicyArn,
      subnetMappings: this.subnetMapping,
      vpcId: props.vpcId,
      deleteProtection: props.deleteProtection,
      description: props.description,
      firewallPolicyChangeProtection: props.firewallPolicyChangeProtection,
      subnetChangeProtection: props.subnetChangeProtection,
      tags: props.tags,
    });

    // Set remaining properties
    this.endpointIds = resource.attrEndpointIds;
    this.firewallArn = resource.ref;
    this.firewallId = resource.attrFirewallId;
  }
}
