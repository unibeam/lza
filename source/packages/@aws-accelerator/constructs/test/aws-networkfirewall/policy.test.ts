/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { FirewallPolicyProperty, NetworkFirewallPolicy } from '../../lib/aws-networkfirewall/policy';
import { snapShotTest } from '../snapshot-test';
import { describe, it } from '@jest/globals';

const testNamePrefix = 'Construct(NetworkFirewallPolicy): ';

const importedFirewallPolicyArn = 'arn:aws:network-firewall:us-east-1:222222222222:firewall-policy/TestImportedPolicy';

//Initialize stack for resource configuration test
const stack = new cdk.Stack();

/**
 * Network Firewall construct test
 */
describe('Network Firewall Policy', () => {
  it('test stateful engine', () => {
    const firewallPolicy: FirewallPolicyProperty = {
      statelessDefaultActions: ['aws:forward_to_sfe'],
      statelessFragmentDefaultActions: ['aws:forward_to_sfe'],
      statefulEngineOptions: 'STRICT_ORDER',
      statefulRuleGroupReferences: [
        {
          priority: 123,
          resourceArn: 'arn:aws:network-firewall:us-east-1:222222222222:stateful-rulegroup/TestGroup',
        },
      ],
    };

    new NetworkFirewallPolicy(stack, 'TestPolicy', {
      firewallPolicy: firewallPolicy,
      name: 'TestFirewallPolicy',
      tags: [],
    });
  });

  it('test custom action', () => {
    const testFirewallPolicy: FirewallPolicyProperty = {
      statelessDefaultActions: ['statelessDefaultActions'],
      statelessFragmentDefaultActions: ['statelessFragmentDefaultActions'],

      statefulDefaultActions: ['statefulDefaultActions'],
      statefulEngineOptions: 'statefulEngineOptions',
      statefulRuleGroupReferences: [
        {
          resourceArn: 'resourceArn',
          priority: 123,
        },
      ],
      statelessCustomActions: [
        {
          actionDefinition: {
            publishMetricAction: {
              dimensions: ['CustomValue'],
            },
          },
          actionName: 'actionName',
        },
      ],
      statelessRuleGroupReferences: [
        {
          priority: 123,
          resourceArn: 'resourceArn',
        },
      ],
    };
    new NetworkFirewallPolicy(stack, 'TestPolicy1', {
      firewallPolicy: testFirewallPolicy,
      name: 'TestFirewallPolicy1',
      tags: [],
    });
  });

  it('test import policy', () => {
    NetworkFirewallPolicy.fromAttributes(stack, 'TestImportPolicy', {
      policyName: importedFirewallPolicyArn,
      policyArn: 'importedPolicyArn',
    });
  });

  snapShotTest(testNamePrefix, stack);
});
