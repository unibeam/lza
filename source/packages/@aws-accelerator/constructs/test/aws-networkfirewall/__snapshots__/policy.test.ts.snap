// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Network Firewall Policy Construct(NetworkFirewallPolicy):  Snapshot Test 1`] = `
{
  "Resources": {
    "TestPolicy14DF854C3": {
      "Properties": {
        "FirewallPolicy": {
          "StatefulDefaultActions": [
            "statefulDefaultActions",
          ],
          "StatefulEngineOptions": {
            "RuleOrder": "statefulEngineOptions",
          },
          "StatefulRuleGroupReferences": [
            {
              "Priority": 123,
              "ResourceArn": "resourceArn",
            },
          ],
          "StatelessCustomActions": [
            {
              "ActionDefinition": {
                "PublishMetricAction": {
                  "Dimensions": [
                    {
                      "Value": "CustomValue",
                    },
                  ],
                },
              },
              "ActionName": "actionName",
            },
          ],
          "StatelessDefaultActions": [
            "statelessDefaultActions",
          ],
          "StatelessFragmentDefaultActions": [
            "statelessFragmentDefaultActions",
          ],
          "StatelessRuleGroupReferences": [
            {
              "Priority": 123,
              "ResourceArn": "resourceArn",
            },
          ],
        },
        "FirewallPolicyName": "TestFirewallPolicy1",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestFirewallPolicy1",
          },
        ],
      },
      "Type": "AWS::NetworkFirewall::FirewallPolicy",
    },
    "TestPolicyCC05E598": {
      "Properties": {
        "FirewallPolicy": {
          "StatefulEngineOptions": {
            "RuleOrder": "STRICT_ORDER",
          },
          "StatefulRuleGroupReferences": [
            {
              "Priority": 123,
              "ResourceArn": "arn:aws:network-firewall:us-east-1:222222222222:stateful-rulegroup/TestGroup",
            },
          ],
          "StatelessDefaultActions": [
            "aws:forward_to_sfe",
          ],
          "StatelessFragmentDefaultActions": [
            "aws:forward_to_sfe",
          ],
        },
        "FirewallPolicyName": "TestFirewallPolicy",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestFirewallPolicy",
          },
        ],
      },
      "Type": "AWS::NetworkFirewall::FirewallPolicy",
    },
  },
}
`;
